frodo.js

```diff
@@ -1,133 +1,131 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?